src/mod/gcewing/sg/features/gdo/client/gui/GateControlComponent.java

```diff
@@ -0,0 +1,168 @@
+import net.malisis.ego.gui.component.container.UIContainer;
+import net.malisis.ego.gui.component.decoration.UIImage;
+import net.malisis.ego.gui.component.decoration.UILabel;
+import net.malisis.ego.gui.component.interaction.UIButton;
+import net.malisis.ego.gui.element.Padding;
+import net.malisis.ego.gui.render.GuiIcon;
+import net.malisis.ego.gui.render.GuiTexture;
+import net.minecraft.util.ResourceLocation;
+
+import java.util.Map;
+
+public class GateControlComponent extends UIContainer {
+
+    private SGBaseTE gate;
+    private boolean irisAccess;
+
+    private final Map<String, GuiIcon> gateTextures = Maps.newHashMap();
+
+    public GateControlComponent(SGBaseTE gate, boolean local) {
+        this.gate = gate;
+
+        //window without title
+        setBackground(SGWindow.createBackground(this, null));
+
+        UILabel title = UILabel.builder()
+                .parent(this)
+                .text(local ? "sgcraft.gui.gdo.label.localGateControl" : "sgcraft.gui.gdo.label.remoteGateControl")
+                .textColor(0xFFFFFF)
+                .scale(1.1F)
+                .topCenter(0, 5)
+                .build();
+
+        UIImage gateStatus = UIImage.builder()
+                .parent(this)
+                .centered()
+                .below(title, 5)
+                .size(110, 110)
+                .icon(this::gateStatus)
+                .build();
+
+        UILabel address = UILabel.builder()
+                .parent(this)
+                .centered()
+                .below(gateStatus, 5)
+                .text(this::address)
+                .textColor(0x5555FF)
+                .scale(1.3F)
+                .build();
+
+        UIButton irisControl = UIButton.builder()
+                .parent(this)
+                .leftAligned()
+                .below(address, 5)
+                .size(70, 20)
+                .text(this::irisControlText)
+                .enabled(this::hasIrisAccess)
+                .onClick(this::openCloseIris)
+                .build();
+
+        if (local) {
+            UIButton.builder()
+                    .parent(this)
+                    .rightOf(irisControl, 5)
+                    .below(address, 5)
+                    .size(70, 20)
+                    .text("sgcraft.gui.button.disconnect")
+                    .enabled(this::isGateConnected)
+                    .onClick(this::disconnectGate)
+                    .build();
+        }
+
+    }
+
+    public void setGate(SGBaseTE gate) {
+        this.gate = gate;
+    }
+
+    public void setIrisAccess(boolean irisAccess) {
+        this.irisAccess = irisAccess;
+    }
+
+    public GuiIcon gateStatus() {
+        if (gate == null) {
+            return GuiIcon.NONE;
+        }
+
+        return gateTextures.computeIfAbsent(gateStatusLabel(), s -> GuiIcon.full(new GuiTexture(new ResourceLocation("sgcraft:" + s))));
+    }
+
+    private String gateStatusLabel() {
+        String status = "textures/";
```

Comment:
StringBuilder, you are creating 5 String objects per call here.

Reply:
I started using StringBuilder, but IDEA complained and told me I should just concatenate strings...

Reply:
IDEA is retarded then.

It is relying on JIT to inline and optimize the call away. I prefer being 100% sure it did.